File: java/src/main/java/polyglot/palindrom/Palindrom.java
package polyglot.palindrom;

import static java.util.Objects.isNull;

public class Palindrom {

    public boolean isPalindromIterative(String value) {
        if (isNull(value)) {
            return false;
        } else if (value.length() <= 1) {
            return true;
        }

        char[] chars = value.toCharArray();
        for (int i = 0; i < chars.length / 2; i++) {
            if (chars[i] != chars[chars.length - i - 1]) {
                return false;
            }
        }

        return true;
    }

    public boolean isPalindromRecursive(String value) {
        if (isNull(value)) {
            return false;
        } else if (value.length() <= 1) {
            return true;
        }

        char first = value.charAt(0);
        char last = value.charAt(value.length() - 1);

        return (first == last) && isPalindromRecursive(value.substring(1, value.length() - 1));
    }

    public boolean isPalindromReverse(String value) {
        if (isNull(value)) {
            return false;
        }

        StringBuilder b = new StringBuilder(value);
        return value.equals(b.reverse().toString());
    }

}

File: kotlin/src/main/kotlin/polyglot/palindrom/Palindrom.kt
package polyglot.palindrom

class Palindrom {

    fun isPalindromIterative(value: String): Boolean {
        if (value.length <= 1) {
            return true
        }

        for (i in 0 until value.length / 2) {
            if (value[i] != value[value.length - i - 1]) {
                return false
            }
        }

        return true
    }

    fun isPalindromRecursive(value: String): Boolean {
        if (value.isEmpty() || value.length <= 1) {
            return true
        }

        val first = value[0]
        val last = value[value.length - 1]

        return first == last && isPalindromRecursive(value.substring(1, value.length - 1))
    }

    fun isPalindromReverse(value: String): Boolean {
        return value.reversed() == value
    }
}
